nupic/research/frameworks/pytorch/cl_model_utils.py

```diff
@@ -0,0 +1,205 @@
+    :param scenario: continual learning scenario, one of "task", "domain", or "class"
+    :param num_classes_per_task: number of classes per task
+    :param loader: train dataset loader
+    :param device: device to use ('cpu' or 'cuda')
+    :param criterion: loss function to use
+    :param post_batch_callback: Callback function to be called after every batch with
+                                the following parameters: model, batch_idx
+    """
+    model.eval()
+    total = 0
+
+    # Perform accumulation on device, avoid paying performance cost of .item()
+    loss = torch.tensor(0., device=device)
+    correct = torch.tensor(0, device=device)
+
+    async_gpu = loader.pin_memory
+
+    with torch.no_grad():
+        for batch_idx, (data, target) in enumerate(loader):
+
+            data = data.to(device, non_blocking=async_gpu)
+            target = target.to(device, non_blocking=async_gpu)
+
+            # Compute loss only for "active" output units, and only backpropogate
+            # errors for these units when calling loss.backward()
+            active_classes = get_active_classes(scenario,
+                                                loader.sampler.active_tasks[-1],
+                                                num_classes_per_task)
+            output = model(data)
+            output = output[:, active_classes]
+
+            loss += criterion(output, target, reduction="sum")
+            pred = output.max(1, keepdim=True)[1]
+            correct += pred.eq(target.view_as(pred)).sum()
+            total += len(data)
+
+            if post_batch_callback is not None:
+                post_batch_callback(batch_idx=batch_idx, target=target,
+                                    output=output, pred=pred)
+
+    correct = correct.item()
+    loss = loss.item()
+
+    result = {
+        "total_correct": correct,
+        "total_tested": total,
+        "mean_loss": loss / total if total > 0 else 0,
+        "mean_accuracy": correct / total if total > 0 else 0,
+    }
+
+    return result
+
+
+def get_active_classes(scenario, task_num, num_classes_per_task):
+    """
+    Returns a list of label indices that are "active" during training under the
+    continual learning scenario specified by the config. In the "task" scenario, only
+    the classes that are being trained are active. In the "class" scenario, all tasks
+    that the model has previously observed are active. More information about active
+    classes can be found here: https://arxiv.org/abs/1904.07734
+
+    :param task_num: zero-based index of the task
+    :param scenario: continual learning scenario, one of "task", "domain", or "class"
+    :param num_classes_per_task: the number of classes per task
+    """
+    if scenario == "task":
+        return [label for label in range(
+            num_classes_per_task * task_num,
+            num_classes_per_task * (task_num + 1)
+        )]
+
+    elif scenario == "domain":
```

Comment:
If we are not implementing, better to remove this. Doesn't seem like we will be working with the domain scenario anytime soon.

Reply:
Ok, I will remove it.